presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -0,0 +1,159 @@
+            }
+            if (!stringBuilder.toString().isEmpty() && !stringBuilder.toString().equals("")) {
+                clusteringPredicates.add(stringBuilder.toString());
+            }
+        }
+        return clusteringPredicates.build();
+    }
+
+    public TupleDomain<ColumnHandle> getUnenforcedConstraints()
+    {
+        Map<ColumnHandle, Domain> pushedDown = clusteringPushDownResult.getDomains();
+        Map<ColumnHandle, Domain> notPushedDown = new HashMap<>(predicates.getDomains().get());
+
+        if (!notPushedDown.isEmpty() && !pushedDown.isEmpty()) {
+            notPushedDown.entrySet().removeAll(pushedDown.entrySet());
+        }
+
+        return TupleDomain.withColumnDomains(notPushedDown);
+    }
+
+    private static ClusteringPushDownResult getClusteringKeysSet(List<CassandraColumnHandle> clusteringColumns, TupleDomain<ColumnHandle> predicates)
+    {
+        ImmutableMap.Builder<ColumnHandle, Domain> domainsBuilder = ImmutableMap.builder();
+        ImmutableList.Builder<Set<Object>> clusteringColumnValues = ImmutableList.builder();
+        for (CassandraColumnHandle columnHandle : clusteringColumns) {
+            Domain domain = predicates.getDomains().get().get(columnHandle);
+
+            if (domain == null) {
+                break;
+            }
+
+            if (domain.isNullAllowed()) {
+                return new ClusteringPushDownResult(domainsBuilder.build(), ImmutableSet.of());
+            }
+
+            Set<Object> values = domain.getValues().getValuesProcessor().transform(
+                    ranges -> {
+                        ImmutableSet.Builder<Object> columnValues = ImmutableSet.builder();
+                        for (Range range : ranges.getOrderedRanges()) {
+                            if (!range.isSingleValue()) {
+                                return ImmutableSet.of();
+                            }
+                            /* TODO add code to handle a range of values for the last column
+                             * Prior to Cassandra 2.2, only the last clustering column can have a range of values
+                             * Take a look at how this is done in PreparedStatementBuilder.java
+                             */
+
+                            Object value = range.getSingleValue();
+
+                            CassandraType valueType = columnHandle.getCassandraType();
+                            columnValues.add(valueType.validateClusteringKey(value));
+                        }
+                        return columnValues.build();
+                    },
+                    discreteValues -> {
+                        if (discreteValues.isWhiteList()) {
+                            return ImmutableSet.copyOf(discreteValues.getValues());
+                        }
+                        return ImmutableSet.of();
+                    },
+                    allOrNone -> ImmutableSet.of());
+
+            clusteringColumnValues.add(values);
```

Comment:
if `values` is empty, does it make sense to add it to the `clusteringColumnValues`? shouldn't it be within the `if` clause?